### PR TITLE
Help request notifications/fixed

### PR DESCRIPTION
**Pull request by 98ZhaoJeffrey**
## Notion ticket link

[Ticket Name](https://www.notion.so/uwblueprintexecs/7e51fde5f22d448b8226f2cfa4dd93d5?v=b2dc0dba06574651a8b6063bdd0d74e0&p=f9fd876634c4411691742f6e7790da71&pm=s)

large-pr

## Implementation description
* Implemented notifications and help request table
* Notifications:
  1) Allows all users (currently only notifies facilitators) to receive real-time notifications
  2) Displays a notification bell that signifies how many unseen notifications there are and goes away after closing the bell
  3) Notification menu is paginated to load 10 at a time (doesn't actually but it should)
* Help requests:
  1) Allow learners to create help requests that notifies their facilitator
  2) Displays everything in a table sorted by time created and provides access to the corresponding module/unit/page
  3) Each help request can be marked as completed/uncompleted and clicking on id brings to the individual page

## Steps to test

I might add a video about doing this cuz these steps are too hard for me to understand too ngl

1. You will need a learner account and a facilitator user where they reference each other in the respective field
2. Open two browser instances and login both accounts. 
3. On the facilitator side, go to /help-requests (you actually need to do this step before step 4)
4. On the learner side go to /ask-for-help and click on the modal and type a request
5. On facilitator side (for steps 5-10), you should see that you get a notification and click on the bell icon will display the message
6. Upon closing the bell, the red circle with the number should go away
7. Click on the notification in the menu and it should take you to /help-requests/:id where id is the help request id. You should see the message and the id of the request
8. Change the id to one that doesn't exist in the database and it should say help request doesn't exist or something
9. Go back to /help-requests by clicking on the link in the top left and toggle the help request as completed. Refresh the page to see if it persists.
10. Make sure that click on the help request id brings you to the same page as the notification too

## What should reviewers focus on?
* ngl this is too long so whoever is reviewing this have fun. Most of the bad code is on the backend though
* I am prolly very inconsistent so like tell me where i should change variable names and stuff

* In notification router, there is a /test route. That shouldn't exist but it's only there since I haven't figured out how to fix an error related to importing mongoose models in the correct order. Dont worry abt testing that. 

2l7u3sqgemo21

## Checklist
- [ ] My PR name is descriptive and in imperative tense
- [ ] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [ ] I have run the appropriate linter(s)
- [ ] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

**Comment by carolynzhang18** [comment]
I'm getting this console error at `/help-requests`:
image

Fix should just be `const [page, setPage] = useState(0);` in `ViewHelpRequestsPage.tsx` :)

File: backend/services/implementations/notificationService.ts [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    // this is digustingly bad ngl
+    const message = updateNotification.message
+      ? updateNotification.message
+      : oldNotification.message;
+    const read = updateNotification.read
+      ? updateNotification.read
+      : oldNotification.read;
+    const link = updateNotification.link
+      ? updateNotification.link
+      : oldNotification.link;
```
Comment (carolynzhang18) [inline backend/services/implementations/notificationService.ts:121]: lol the best I can come up with is
```
const message = updateNotification.message ?? oldNotification.message;
const read = updateNotification.read ?? oldNotification.read;
const link = updateNotification.link ?? oldNotification.link;
```

Comment (carolynzhang18) [reply backend/services/implementations/notificationService.ts:121]: the other idea is returning 
```
return {
    ...oldNotification.
    ...updateNotification
};
```
which shoooould give the fields of oldNotification overwritten by whichever fields of updateNotification exist...?

File: backend/services/implementations/notificationService.ts [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+          read: true,
+        },
+      );
+    } catch (error) {
+      Logger.error(
+        `Failed to update notifications. Reason = ${getErrorMessage(error)}`,
+      );
+      throw error;
+    }
+    return updates;
```
Comment (carolynzhang18) [inline backend/services/implementations/notificationService.ts:147]: does this return the updated documents? or the documents as they were found, before the updates?

File: frontend/src/components/pages/HelpRequestPage.tsx [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+      </Button>
+      {helpRequest ? (
+        <>
+          Help request ID: {id}
+          <div>{helpRequest?.message}</div>
+        </>
+      ) : (
+        <div>Help request with ID: {id} does not exist</div>
+      )}
+    </div>
```
[the remark this answers is not in the document]
Comment (98ZhaoJeffrey) [reply frontend/src/components/pages/HelpRequestPage.tsx:42]: i do a bit of trolling, yeah thats prolly correct

File: backend/services/implementations/helpRequestService.ts [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    }
+    return {
+      id: oldHelpRequest.id,
+      message: oldHelpRequest.message,
+      learner: oldHelpRequest.learner,
+      facilitator: oldHelpRequest.facilitator,
+      unit: updates.unit !== undefined ? updates.unit : oldHelpRequest.unit,
+      module:
+        updates.module !== undefined ? updates.module : oldHelpRequest.module,
+      page: updates.page !== undefined ? updates.page : oldHelpRequest.page,
```
[the remark this answers is not in the document]
Comment (98ZhaoJeffrey) [reply backend/services/implementations/helpRequestService.ts:96]: honestly i forgot how ts worked when doing this part

File: backend/services/implementations/notificationService.ts (hunk 1 above)
Comment (98ZhaoJeffrey) [reply backend/services/implementations/notificationService.ts:121]: I swear I tried this and it was giving me the wrong result. Maybe I remembered wrong cuz that's why I didn't do it this way

File: backend/rest/notificationRoutes.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  const { user, skip, limit } = req.query;
+  try {
+    if (!user) {
+      res.status(400).send("User id must be provided");
+      return;
+    }
+    const data = await notificationService.getNotifications(
+      user as string,
+      parseInt(skip as string, 10),
+      parseInt(limit as string, 10),
```
[the remark this answers is not in the document]
Comment (98ZhaoJeffrey) [reply backend/rest/notificationRoutes.ts:20]: OK if it makes this part less ugly to work with, I can change to body parameters. Idt there was a big reason picking body vs search params

File: backend/services/implementations/notificationService.ts (hunk 2 above)
Comment (98ZhaoJeffrey) [reply backend/services/implementations/notificationService.ts:147]: I think all 3 actually + 1 other field

File: backend/models/helprequest.mgmodel.ts [hunk 6]
```diff
@@ -17,27 +19,46 @@ const HelpRequestSchema: Schema = new Schema(
       required: true,
     },
     learner: {
-      type: String,
+      type: mongoose.Schema.Types.ObjectId,
```
[the remark this answers is not in the document]
Comment (98ZhaoJeffrey) [reply backend/models/helprequest.mgmodel.ts:22]: do u mean just Schema.Types.ObjectId cuz just ObjectId doesn't seem to work for me

**Comment by 98ZhaoJeffrey** [comment]
ngl i dont what is exactly the problem is but i do get rate limited on help request page and just going between individual help request page and the view all help request page (which is a likely use case). Maybe some caching is needed?

**Review by carolynzhang18 — APPROVED** [review]

**Pull request merged**